annotated.go

```diff
@@ -195,6 +196,310 @@ func ResultTags(tags ...string) Annotation {
+		},
+	}
+
+	type valueResolver func(reflect.Value, int) reflect.Value
+	type argSource struct {
+		pos     int
+		result  bool
+		resolve valueResolver
+	}
+
+	resolverIdx := make([]argSource, 1)
+	ft := la.targetType()
+	for i := 1; i < ft.NumIn(); i++ {
+		t := ft.In(i)
+		resultIdx, isProvidedByResults := resultMap[t]
+
+		if isProvidedByResults {
+			resolverIdx = append(resolverIdx, argSource{
+				pos:     i,
+				result:  true,
+				resolve: resultIdx.resolve,
+			})
+			continue
+		}
+
+		field := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: t,
+		}
+		params = append(params, field)
+
+		resolver := func(v reflect.Value, pos int) (value reflect.Value) {
+			value = la.resolveLifecycleParamField(v, i)
+			return
+		}
+
+		resolverIdx = append(resolverIdx, argSource{
+			pos:     i,
+			resolve: resolver,
+		})
+	}
+
+	in = reflect.StructOf(params)
+
+	argmap = func(
+		args []reflect.Value,
+	) (lc Lifecycle, remapped []reflect.Value) {
+		remapped = make([]reflect.Value, ft.NumIn())
+
+		if len(args) != 0 {
+
+			p := args[0]
+			results := args[1]
+
+			lc, _ = p.FieldByName("Lifecycle").Interface().(Lifecycle)
+
+			for i := 1; i < ft.NumIn(); i++ {
+				resolver := resolverIdx[i]
+
+				source := p
+				if resolver.result {
+					source = results
+				}
+
+				remapped[i] = resolver.resolve(source, i)
+			}
+		}
+
+		return
+	}
+	return
+}
+
+func (la *lifecycleHookAnnotation) buildHook(fn func(context.Context) error) (hook Hook) {
+	switch la.Type {
+	case _onStartHookType:
+		hook.OnStart = fn
+	case _onStopHookType:
+		hook.OnStop = fn
+	}
+
+	return
+}
+
+func (la *lifecycleHookAnnotation) Build(results ...reflect.Type) (reflect.Value, error) {
+	in, paramMap := la.parameters(results...)
+	params := []reflect.Type{in}
+	for _, r := range results {
+		if r != _typeOfError {
+			params = append(params, r)
+		}
+	}
+
+	origFn := reflect.ValueOf(la.Target)
+	newFnType := reflect.FuncOf(params, nil, false)
+	newFn := reflect.MakeFunc(newFnType, func(args []reflect.Value) []reflect.Value {
+		var lc Lifecycle
+		lc, args = paramMap(args)
+		hookFn := func(ctx context.Context) (err error) {
+			args[0] = reflect.ValueOf(ctx)
+
+			results := origFn.Call(args)
+			if len(results) > 0 && results[0].Type() == _typeOfError {
+				err, _ = results[0].Interface().(error)
+			}
+
+			return
+		}
+
+		lc.Append(la.buildHook(hookFn))
+		return []reflect.Value{}
+	})
+
+	return newFn, nil
+}
+
+// OnStart is an Annotation that appends an OnStart Hook to the application
+// Lifecycle when that function is called. This provides a way to create
+// Lifecycle OnStart hooks without building a function that takes a dependency
+// on the Lifecycle type.
+//
+//  fx.Annotate(
+//    func(...) Server { ... },
+//    fx.OnStart(func(ctx context.Context, server Server) error {
+//        return server.Listen(ctx)
+//    }),
+//  )
+//
+// Only one OnStart annotation may be applied to a given function at a time.
```

Comment:
Let's add a clarification on this - we can provide OnStart/OnStop annotations to the same function at the same time. Just not two OnStarts.